amundsen_application/static/js/components/SearchPage/index.tsx

```diff
@@ -93,6 +93,15 @@ export class SearchPage extends React.Component<SearchPageProps> {
     const nextUrlParams = this.getUrlParams(this.props.location.search);
     const prevUrlParams = this.getUrlParams(prevProps.location.search);
 
+    if (!this.props.isLoading && prevProps.isLoading) {
```

Comment:
`componentDidUpdate` fires repeatedly every time any of the props or URL state changes. This tries to capture the moment when a search request completes.

In the case of a SearchAll request, the page URL is set to `search?searchTerm=X`, so the `pageIndex` and `selectedTab` are null. Once the search is complete and a `pageIndex` and `selectedTab` are set in the `sagas` file, we need to update the URL so we don't trigger another update in the next cycle.

This is the cornerstone piece of logic that blocks the myriad of bugs that occur when application state and page URL get out of sync.

Reply:
I think much of this complexity is an inherent property of the number of cases that we handle, and for the most part will be unavoidable for the code to reflect that to some extent.

1. Execute search from non-search page (componentDidMount)
2. Execute search from search page (componentDidUpdate)
3. Re-execute search from search bar with same search term (reset tab, page index, filters)
4. Execute search from external link or page refresh
5. Re-execute/load search results from navigation (back/forward)
6. Re-execute/load search results from `<- Search Results` button
7. Update URL on tab change
8. Update URL from page index change
9. (new) auto-select tab when results are empty for default tab

In addition to handling these actions, we use the URL to read and write the application state at all times. I think we can simplify all of this logic by limiting when and where we read/write to the page URL. For example we could only read the URL on `componentDidMount`, exclusively use Redux actions to update application state, and write the state to URL after each action after each action. I think this could be as a follow up task to this, or before we implement filters.

Reply:
In that example, `SearchBar` would call `this.props.searchAll` and clicking the tab would call `this.props.updateSearchTab`. `SearchPage` would then only need listen to updated props via global search state. 

We could build a separate component that reads the search params and translates them into global state updates and serve as an interface between URL and application state.